pennylane/optimize/rotosolve.py

```diff
@@ -14,78 +14,122 @@
+    ``num_steps`` times. Signature is as expected by ``RotosolveOptimizer._min_numeric``
+    below, returning a scalar minimal position and the function value at that position."""
     width = 2 * np.pi
     x_min = 0.0
     Ns = kwargs.pop("Ns")
     for _ in range(num_steps):
         _range = (x_min - width / 2, x_min + width / 2)
         x_min, y_min, *_ = brute(fun, ranges=(_range,), full_output=True, Ns=Ns, **kwargs)
+        # We only ever use this function for 1D optimization
+        x_min = x_min[0]
         width /= Ns
 
     return x_min, y_min
 
 
 def _shgo_optimizer(fun, **kwargs):
     r"""Wrapper for ``scipy.optimize.shgo`` (Simplicial Homology global optimizer).
-    Signature is as expected by ``RotosolveOptimizer._rotosolve`` below, providing
+    Signature is as expected by ``RotosolveOptimizer._min_numeric`` below, providing
     a scalar minimal position and the function value at that position."""
     opt_res = shgo(fun, **kwargs)
-    return opt_res.x, opt_res.fun
+    return opt_res.x[0], opt_res.fun
+
+
+def _prepare_optimizer(optimizer, optimizer_kwargs):
+    r"""Set default optimizer and optimizer keyword arguments
+    for the one-dimensional optimization in each substep of Rotosolve."""
+    optimizer_kwargs = optimizer_kwargs or {}
+    optimizer = optimizer or "brute"
+
+    if optimizer == "brute":
+        optimizer = _brute_optimizer
+        optimizer_kwargs.setdefault("num_steps", 4)
+        optimizer_kwargs.setdefault("Ns", 100)
+    elif optimizer == "shgo":
+        optimizer = _shgo_optimizer
+        optimizer_kwargs.setdefault("bounds", ((-np.pi, np.pi),))
+
+    return optimizer, optimizer_kwargs
+
+
+def _restrict_to_univariate(fn, arg_idx, par_idx, args, kwargs):
+    r"""Restrict a function to a univariate function for given argument
+    and parameter indices.
+
+    Args:
+        fn (callable): Multivariate function
+        arg_idx (int): Index of the argument that contains the parameter to restrict to
+        par_idx (tuple[int]): Index of the parameter to restrict to within the argument
+        args (tuple): Arguments at which to restrict the function.
+        kwargs (dict): Keyword arguments at which to restrict the function.
+
+    Returns:
+        callable: Univariate restriction of ``fn``. That is, this callable takes
+        a single float value as input and has the same return type as ``fn``.
+        All arguments are set to the given ``args`` and the input value to this
+        function is added to the marked parameter.
+    """
+    the_arg = args[arg_idx]
+    if len(qml.math.shape(the_arg)) == 0:
+        shift_vec = qml.math.ones_like(the_arg)
+    else:
+        shift_vec = qml.math.zeros_like(the_arg)
+        shift_vec = qml.math.scatter_element_add(shift_vec, par_idx, 1.0)
+
+    def _univariate_fn(x):
+        return fn(*args[:arg_idx], the_arg + shift_vec * x, *args[arg_idx + 1 :], **kwargs)
```

Comment:
out of curiosity, does this result in any gotchas/restrictions as to the form of the cost function signature?

Reply:
I believe that no, but it's hard to be sure on such a question :sweat_smile: 
The signature is "restricted" to be `*args, **kwargs` and those `args` that are passed values with `requires_grad=True` are expected to be floats or arrays.

Reply:
okay, let's leave it as is, and just keep an eye out on potential user reported bugs if there are edge cases